app/services/authentication/providers/provider.rb

```diff
@@ -0,0 +1,60 @@
+module Authentication
+  module Providers
+    # Authentication provider
+    class Provider
+      # name of the field to store the upstream identity's creation date in
+      CREATED_AT_FIELD = "".freeze
+      # name of the field to store the upstream identity's username in
+      USERNAME_FIELD = "".freeze
+
+      def initialize(auth_payload)
+        @auth_payload = cleanup_payload(auth_payload.dup)
+        @info = auth_payload.info
+        @raw_info = auth_payload.extra.raw_info
+      end
+
+      # Extract data for a brand new user
+      def new_user_data
+        raise NotImplementedError
```

Comment:
Unlike Python's `NotImplementedError` Ruby's is for platform/OS-specific things:

> Raised when a feature is not implemented on the current platform. For example, methods depending on the `fsync` or `fork` system calls may raise this exception if the underlying operating system or Ruby runtime does not support them.

Because of this, it's considered non-rescuable and therefore doesn't inherit `StandardError` but `Exception` (via `ScriptError`):

```ruby
begin
  Authentication::Providers::Provider.new({})
rescue StandardError => _e
  puts "All good"
end
#=> NotImplementedError: NotImplementedError
```

I generally add a Smalltalk inspired `SubclassResponsibility` exception to my global error definitions (`app/errors` in our case).

Reply:
OMG, totally didn't know this, I thought it was the same ahhaha. Fixed it and added `SubclassResponsibility`

Reply:
Haha, yeah, it's confusing, especially since they have exactly the same name. Thanks for adding the new exception!